src/porepy/numerics/ad/operator_functions.py

```diff
@@ -0,0 +1,617 @@
+            # it returns None-like things for every iteration more then shortest length
+            # These Nones are ignored for some reason by the function call, as well as by the
+            # array constructor.
+            # If a mortar var and a subdomain var are given as args,
+            # then the lengths will be different for example.
+            return np.array([self.func(*vals_i) for vals_i in zip(*vals)])
+
+
+### CONCRETE IMPLEMENTATIONS ------------------------------------------------------------------
+
+
+class Function(AbstractFunction):
+    """Ad representation of an analytically given function,
+    where it is expected that passing Ad_arrays directly to ``func`` will
+    return the proper result.
+
+    Here the values AND the Jacobian are obtained exactly by the AD framework.
+
+    The intended use is as a wrapper for operations on pp.ad.Ad_array objects,
+    in forms which are not directly or easily expressed by the rest of the Ad
+    framework.
+
+    NOTE: This is a special case where the abstract methods for getting values and the Jacobian
+    are formally implemented but never used by the AD framework (see ``ad_compatible``)
+
+    """
+
+    def __init__(self, func: Callable, name: str, array_compatible: bool = True):
+        super().__init__(func, name, array_compatible, True)
+
+    def get_values(self, *args: Ad_array) -> np.ndarray:
+        result = self.func(*args)
+        return result.val
+
+    def get_jacobian(self, *args: Ad_array) -> np.ndarray:
+        result = self.func(*args)
+        return result.jac
+
+
+class ConstantFunction(AbstractFunction):
+    """ "Function representing constant, scalar values with no dependencies and ergo a
+    zero Jacobian.
+
+    It still has to be called though since it fulfills the notion of a 'function'.
+    """
+
+    def __init__(self, name: str, values: np.ndarray):
+        """
+        Args:
+            values: constant values per cell
+
+        """
+        # dummy function, takes whatever and returns only the pre-set values
+        def func(*args):
+            return values
+
+        super().__init__(func, name)
+        self._values = values
+
+    def get_values(self, *args: Ad_array) -> np.ndarray:
+        """
+        Returns:
+            The values passed at instantiation.
+
+        """
+        return self._values
+
+    def get_jacobian(self, *args: Ad_array) -> sps.spmatrix:
+        """
+        IMPLEMENTATION NOTE:
+        The return value is not a sparse matrix as imposed by the parent method signature,
+        but a zero.
+        Numerical operations with a zero always works with any numeric formats in
+        numpy, scipy and PorePy's AD framework.
+        Since the constant function (most likely) gets no arguments passed,
+        we have no way of knowing the necessary SHAPE for a zero matrix. Hence scalar.
+
+        Returns:
+            The trivial derivative of a constant.
+
+        """
+        return 0.0
```

Comment:
@keileg  This unfortunately will have to stay, as explained in the implementation note.

Reply:
I figured that might be a problem. Let's keep this for now, but mypy will likely force us to reconsider at some point.